Merge special fields

scgao opened the pull request:
I implemented insertId/operation/sourceLocation and merge them into one file, `stackdriver_special_fields`. 
Besides, I remove the test codes.

After I pass the first PR, I will send this PR.

Welcome to suggest!

scgao committed:
out_stackdriver: added support for operation field

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update cmake file - add stackdriver_operation.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests:runtime:out_stackdriver: unit tests for adding 'operation' field
in stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
tests: add out_stackdriver.c

Signed-off-by: scgao <scgao@google.com>

scgao committed:
config: add customized config file for stackdriver

Signed-off-by: scgao <scgao@google.com>

scgao committed:
deal with memory lead

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove local config files

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: clean up id/producer inside the while loop

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update based on the code review
In extrack_operation(), use 2 different keys, one for operation and the
other for logging.googleapis.com/operation;
Deal with memory leak inside extrack_operation;

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'operation_existed' as 'operation_extracted

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: rename 'update_json_payload()' as 'pack_json_payload'

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: deal with the heap leak - no heap leak now

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove calculate_spec_fields(), calulate the entry size inline

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
reconstruct the operation_status inside extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: construct operation_status to present whether the operation field is valid

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: Design several test cases and do runtime test

Signed-off-by: scgao <scgao@google.com>

scgao committed:
test: runtime: unit tests for stackdriver operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: remove trouble_shooting code

Signed-off-by: scgao <scgao@google.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
Update plugins/out_stackdriver/stackdriver_operation.c

Co-authored-by: Yen-Cheng Chou <3055688+StevenYCChou@users.noreply.github.com>

scgao committed:
out_stackdriver: remove operation_status to stackdriver_opration.c and
simplify the return logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the logic of extrack_operation()

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: simplify the login in extract_operation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: use flb_sds_copy() to change the value of flb_sds_t

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: support 'insertId' field

scgao committed:
out_stackdriver: implement the second feature gap - insertId

scgao committed:
out_stackdriver: update pack_json_payload(), thus can deal with multiple special fields in the jsonPayload

scgao committed:
out_stackdriver: update logic of extract_operation()

scgao committed:
out_stackdriver: update the logic of extract_operation()

scgao committed:
out_stackdriver: simplify the logic of extract_operation()

scgao committed:
Merge branch 'update_review_2' into add_insertId

scgao committed:
outstackdriver: support sourceLocation

scgao committed:
out_stackdriver: update sub-field of operation field. Make them optional.

scgao committed:
Merge branch 'update_sub_field' into add_sourceLocation

scgao committed:
out_stackdriver: support sourceLocation

scgao committed:
out_stackdriver: sourceLocation v1 - haven't dealed with the type of 'line'

scgao committed:
out_stackdriver: complete sourceLocation v1

scgao committed:
out_stackdriver: finish supporting sourceLocation

scgao committed:
Merge branch 'master' into add_sourceLocation

scgao committed:
Merge branch 'master' into add_sourceLocation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update sourceLocation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update sourceLocation

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update insertId

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update insertId/sourceLocation
Extract insertId as an object

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: update CMakeList

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: extract sourceLocation as an object;
Deal with different cases of 'line'.

Signed-off-by: scgao <scgao@google.com>

scgao committed:
out_stackdriver: merge insertId/operation/sourceLocation into one file

Signed-off-by: scgao <scgao@google.com>

StevenYCChou commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -444,16 +444,47 @@ static int get_severity_level(severity_t * s, const msgpack_object * o,
+        for(; kv != kvend; ++kv) {
+            cur_key = flb_sds_copy(cur_key, kv->key.via.str.ptr, kv->key.via.str.size);
+
+            if (insertId_extracted && strcmp(cur_key, "insertId") == 0 
```
extract `insertID` as a constant as well

scgao replied:
Done.

StevenYCChou commented on plugins/out_stackdriver/stackdriver.c:
```diff
@@ -591,17 +646,33 @@ static int stackdriver_format(const void *data, size_t bytes,
 
+        /* Add sourceLocation field into the log entry */
+        if (sourceLocation_extracted) {
+            /*add_sourceLocation_field(&sourceLocation_file, &sourceLocation_line, 
```
remove unused comments

scgao replied:
Done.

StevenYCChou commented on plugins/out_stackdriver/stackdriver_special_fields.c:
```diff
@@ -67,32 +67,73 @@ bool extract_operation(msgpack_object *operation, msgpack_object *obj)
+    int i = 0;
+
+    for(; i < size; ++ i) {
+        if(s[i] > '9' || s[i] < '0') {
```
Use c's library `isdigit()` from <ctype.h>

scgao replied:
Done.

StevenYCChou commented on plugins/out_stackdriver/stackdriver_special_fields.c:
```diff
@@ -67,32 +67,73 @@ bool extract_operation(msgpack_object *operation, msgpack_object *obj)
+                    }
+                    else if (strncmp("line", tmp_p->key.via.str.ptr, tmp_p->key.via.str.size) == 0
+                             && tmp_p->val.type == MSGPACK_OBJECT_POSITIVE_INTEGER) {
+                        continue;
```
Do you need to update the line to string if it's a valid `line` so the backend would accept it?

scgao replied:
The backend will accept both valid `string` and `int_64`. We can simulate interacting with the backend API on: https://cloud.google.com/logging/docs/reference/v2/rest/v2/entries/write.

Thus a `line` in `int_64` format is good enough.

scgao committed:
out_stackdriver: update after the code review
- remove unused comments
- extract 'insertId' as a constant
- use isdigit() when validating 'line' in 'sourceLocation'

Signed-off-by: scgao <scgao@google.com>
